### PR TITLE
EES-3913 Add free text key stats to the frontend

### PR DESCRIPTION
**Pull request by markhiveit**
This PR adds free text key stat tiles to the frontend. This includes some UI layout changes to key stats on the Admin content page - credit to Marv - visually separating the adding key stat functionality from the existing key stat / editing functionality.

I've made this PR as a draft, as I've not been able to run the UI tests - ~~I'm waiting for [EES-2469](https://github.com/dfe-analytical-services/explore-education-statistics/pull/3897) to be merged~~. Once that is merged, I'll fix the UI tests I've added, which inevitably will be failing. Edit: Oops - accidentally didn't create it as a draft - anyway I will fix regardless

### Other changes:
- Introduced `KeyStatisticType` enum - hopefully makes some key stat code clearer and less error prone.
- Made naming/labelling of key stat "statistic" consistent across the codebase - This is actually a problem of my own creation. The key stat "statistic" was originally named "value", but with the new key stat work I used  "statistic". With hindsight, I would have made my naming consistent with prior naming and used "value", but since it was 50/50 when I realised the issue, I've decided to go with "statistic".

### CODE REVIEW

File: src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx
Comment (amyb-hiveit) [inline src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx:133]: This seems like a fairly large component to have in the same file, maybe better in it's own one?

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (amyb-hiveit) [inline src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:176]: This test seems wrong, it's testing the preview version, which is covered in other tests, rather than that clicking edit renders the form with populated inputs.

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (amyb-hiveit) [inline src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:181]: Generally we try to not have a lot in the waitFors and to especially avoid getByRole as this causes slowness running the tests. You can just wait for one thing then continue to check the rest after it.

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (amyb-hiveit) [inline src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:250]: Rather than have this as a comment it'd be better to have a specific test for trimming whitespace.

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (amyb-hiveit) [inline src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:238]: Don't know if we need to await all these userEvents

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (amyb-hiveit) [inline src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:279]: Do we need this check?

File: src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx
Comment (luke-h1) [inline src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx:186]: Don't think we need the cast here?

File: src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx
Comment (luke-h1) [inline src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx:195]: No need to return a fragment here afaik

File: src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx:49]: Good catch

File: src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx:20]: Seems I thought this was required because it was used in two different components, but I overlooked that they're both in the same file somehow.

Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx:20]: I've left this as exported now as I'm going to move AddKeyStatistics to its own file.

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:250]: Also added the same additional test to `EditableKeyStatDataBlock.test.tsx`

File: src/explore-education-statistics-common/src/services/publicationService.ts
Comment (markhiveit) [reply src/explore-education-statistics-common/src/services/publicationService.ts:97]: I'll let Amy or Nick chime in on this and go along with whatever you lot think is best.

File: tests/robot-tests/tests/admin/bau/release_content.robot
Comment (markhiveit) [reply tests/robot-tests/tests/admin/bau/release_content.robot:173]: The autocomplete in pycharm suddenly started capitalising, and I couldn't be bothered to change it. I've changed it now, and all other instances where it is capitalised.

File: src/explore-education-statistics-admin/src/services/keyStatisticService.ts
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/services/keyStatisticService.ts:29]: On the backend, we've been making each their own object - this means that responses and requests have to be deliberately altered. I think this more to prevent mischief on the response side - i.e. someone makes a change that unwittingly means an endpoint suddenly starts returning something we didn't intend. 

Not sure there is as much danger on the request side - someone could pass in something we didn't intend, but then it wouldn't get used, I think. Maybe there is some danger in some edge case.

I'm inclined to keep them separate - if only because that's how they are in the backend. But if others think differently, happy to go along with something else.

File: src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx:134]: I struggled with the name of this. Can rename to something else if you've got something better in mind.

File: src/explore-education-statistics-common/src/services/publicationService.ts
Comment (markhiveit) [inline src/explore-education-statistics-common/src/services/publicationService.ts:96]: On an aside, I've added this no-shadow disable, copying a different place that we used an enum in the codebase, but I think we can change the eslint config so these aren't necessary:
https://typescript-eslint.io/rules/no-shadow/
```
module.exports = {
  "rules": {
    // Note: you must disable the base rule as it can report incorrect errors
    "no-shadow": "off",
    "@typescript-eslint/no-shadow": "warn"
  }
};
```
Not sure if this is right or not, so I've followed the established way.

File: src/explore-education-statistics-admin/src/services/keyStatisticService.ts
Comment (luke-h1) [reply src/explore-education-statistics-admin/src/services/keyStatisticService.ts:29]: In the future, you could just omit properties that don't belong in the update request by extending from the create request. My only gripe with you doing it this way is you're repeating yourself in two interfaces & will have to update both if anything about both change

File: src/explore-education-statistics-common/src/services/publicationService.ts
Comment (amyb-hiveit) [reply src/explore-education-statistics-common/src/services/publicationService.ts:97]: I think Luke's right here, there is one instance of using an enum in the codebase but generally we'd use the pattern he describes above.

Comment (ntsim) [reply src/explore-education-statistics-common/src/services/publicationService.ts:97]: Yep, I wouldn't recommend TypeScript enums as they're generally considered a bit of a mistake by the core developers (they wouldn't have put them in if they were to start over).

File: src/explore-education-statistics-admin/src/services/keyStatisticService.ts
Comment (ntsim) [reply src/explore-education-statistics-admin/src/services/keyStatisticService.ts:29]: I think across codebase you'll be able to find examples of us duplicating models or not, depending on the scenario. Some cases the update will differ from the create.

However, I probably would suggest that we could create single request view model i.e. `KeyStatisticTextSaveRequest`, then inherit from it with the create/update requests i.e.

```cs
public record KeyStatisticTextSaveRequest {
  // the fields...
}

public record KeyStatisticTextCreateRequest : KeyStatisticTextSaveRequest;

public record KeyStatisticTextUpdateRequest : KeyStatisticTextSaveRequest;
```

Typically, with REST APIs, you expect a bit of symmetry when creating/updating i.e. anything you can set in the create can also be done in the update (if it's a PUT).

Of course, this isn't always the case, but I don't think key statistics will really warrant any major discrepancies in the request models right now. We can always refactor it anyway!

If we do make the above change, I'd go with what Luke has suggested and model this in a similar way in the frontend too.

File: tests/robot-tests/tests/admin/bau/release_content.robot
Comment (ntsim) [reply tests/robot-tests/tests/admin/bau/release_content.robot:173]: You should have a 'Capitalize Keywords' options in your Robot options:

image

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (ntsim) [reply src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:181]: The `ByText`, `ByLabel` or `ByTestId` selectors are much much faster and are acceptable to use in the `waitFor`

File: src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.module.scss
Comment (ntsim) [reply src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.module.scss:9]: Yep, don't add additional verbosity if we don't need to. This can just be `formContainer` like it was previously

File: src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.module.scss
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.module.scss:13]: As above, no need to have the `keyStat` prefix. Could just be `textFormContainer`

File: src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx:77]: Discussed this with Amy and Marv. We've arrived at this:
image

This is a compromise between dev concerns and UI concerns. Changing the colour of the `Enter a statistic` error was going to be a bit of a pain, so first suggestion was to get rid of the blue tile background completely. But having done that, I thought it is good for users to have these three field visually grouped somehow, which is how we eventually arrived at the above screenshot.

I did consider changing the styling for the key stat data block edit form, but decided to leave it for now. There is no validation errors possible on that form, so the background can remain blue.

File: src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx
Comment (ntsim) [reply src/explore-education-statistics-admin/src/pages/release/content/components/KeyStatistics.tsx:134]: `setFormType`

File: src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx:30]: Bit of a nitpick - why have we moved these? Although we're a bit inconsistent, we typically place `on*` event handler props last. Depending on the case, we may also alphabetise the rest of the props as well.

File: src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx:38]: As above - why have we moved these?

File: src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx:77]: On the gap between the ErrorSummary and the form proper, I've left that as it is. This is bottom margin on the ErrorSummary and I believe this is set the same across the whole site, and would also be a bit faffy to customise for this particular instance, so I'm inclined to leave it as it is - but can change if people feel strongly about it.

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:176]: Made similar changes to the test in `EditableKeyStatDataBlock.test.tsx`

File: src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx:30]: Honestly can't remember - I've moved them back.

File: src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStatTextForm.tsx:38]: I've moved the back.

File: src/explore-education-statistics-common/src/services/publicationService.ts
Comment (markhiveit) [reply src/explore-education-statistics-common/src/services/publicationService.ts:96]: Talked to Nick - going to add the config so we can remove the disable lines

File: src/explore-education-statistics-admin/src/services/keyStatisticService.ts
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/services/keyStatisticService.ts:29]: I've introduced `KeyStatisticTextSaveRequest` on both the backend and frontend.

File: tests/robot-tests/tests/admin/bau/release_content.robot
Comment (markhiveit) [reply tests/robot-tests/tests/admin/bau/release_content.robot:173]: Didn't even think to check - ta.

File: src/explore-education-statistics-common/src/services/publicationService.ts
Comment (markhiveit) [reply src/explore-education-statistics-common/src/services/publicationService.ts:97]: Discussed with Luke and Nick - we've gone with something similar to Luke's suggestion above.

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (amyb-hiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:226]: Did you see if these awaits are all necessary? VSCode reckons the ones on userEvent.clear aren't:
await

**Comment by markhiveit** [comment]
> Looks good, just one comment about the awaits in a test, but good to approve.

Ah my bad - somehow marked it has done but didn't do it - I think I somehow managed to interpret that message to be about the waitFors...

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx]: Seems VSCode is right - I'll remove them from EditableKeyStatDataBlock.test.tsx too.

File: src/explore-education-statistics-admin/src/pages/release/content/components/AddKeyStatistics.tsx
Comment (luke-h1) [inline src/explore-education-statistics-admin/src/pages/release/content/components/AddKeyStatistics.tsx:57]: Might want to add a testId here given you've added one for `KeyStatisticText`

File: src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStat.tsx
Comment (luke-h1) [inline src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStat.tsx]: Unused import.

File: src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStat.tsx
Comment (luke-h1) [inline src/explore-education-statistics-admin/src/pages/release/content/components/EditableKeyStat.tsx:52]: Could just spread these values to be more succinct: 
```
            const request: KeyStatisticDataBlockUpdateRequest = {
              ...values
            };
```

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatDataBlock.test.tsx
Comment (luke-h1) [inline src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatDataBlock.test.tsx:437]: Wrong style for typing arrays here. You'll struggle to find this syntax used Typescript codebases & this it goes against our current code styles

```
KeyStatDataBlockFormValues[]
```

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (luke-h1) [inline src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:4]: not part of your change but we should optimize this import while you're here (we're currently importing the entire lodash lib here which is big) 

```
import noop from 'lodash/noop';
```

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (luke-h1) [inline src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:247]: Similar to my other comment around interfaces

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (luke-h1) [inline src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:287]: Similar to my other comment around interfaces

File: tests/robot-tests/tests/libs/admin-common.robot
Comment (luke-h1) [inline tests/robot-tests/tests/libs/admin-common.robot:849]: you can use `user clears element text` here instead. I want to get around to swapping out the implementation of user clears element text in the future. The main reason is SeleniumLibrary has a bug which means anything todo with the keyboard doesn't work properly on Mac.  I'm planning on introducing a new library to sort this but it'd be good if we could consolidate keyboard usage such as this into one central place

**Review by luke-h1 — CHANGES_REQUESTED** [review]

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatDataBlock.test.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatDataBlock.test.tsx:437]: Changed similar cases in the same file and in `EditableKeyStatText.test.tsx`

File: src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/release/content/components/__tests__/EditableKeyStatText.test.tsx:4]: Made the same change in `EditableKeyStatDataBlock.test.tsx` and `CancelAmendmentModal.test.tsx`

**Review by luke-h1 — APPROVED** [review]

**Pull request merged**
